Allow metrics to be filterable

trevan opened the pull request:
## Summary

This is an additional fix for elastic/kibana#9120 which makes metric buckets filterable.  I used the aggConfig.isFilterable method to determine if a metric can be filterable or not.  Only the cardinality metric has been enabled as I wasn't sure if this would fit Kibana's direction.

I also made the metric vis to add the metric filter to show it working.  In addition, it looks like the Inspector table automatically picked up filtering on cardinality metrics.  The table vis was partially working, though.  It showed the cell as filterable but nothing happened until I changed VisFiltersProvider.filter to use aggConfig.isFilterable.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

~~- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
~~- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

trevan commented:
@timroes, you did the initial work to make metric visualization clickable (#16758).  Is this change ok?

timroes commented:
Hi, thanks for the PR. Sorry I've missed the ping. I'll forward this to @lukeelmers and @ppisljar. Could you please check if that still makes sense prior to the expressions, and also will continue working then?

ppisljar commented:
i don't have any objections

trevan commented:
Are you ok with the usage of this.props.vis.API.queryFilter.addFilters()?

And I noticed that all the uses aggConfig.isFilterable() also check if the aggConfig has a field or if the field is filterable.  Are you ok with me moving those checks into aggConfig.isFilterable()?  Or maybe creating a new function that wraps that code together?

trevan commented:
I removed the usage of API.queryFilter.addFilters().

I added createFilter to most of the metrics.  I also added an isFilterable to some of the metrics because the parent pipeline metrics can be filterable depending on the field used in them.

I don't know if I can make the sibling pipeline metrics filterable.  Since they use bucket aggregations, I would need to grab the results to create the filters and I would also need to use a boolean filter to OR them together.  I don't see a way to do that with the current code.

I have added a test for one of the new filters.  I'll add some more tests in the next few days.

ppisljar commented:
ok after taking a bit deeper look i might have some objections :)

in general, the concept of being able to filter on the metrics ... that sounds completely ok, but with the introduction of the pipeline: https://github.com/elastic/kibana/pull/25996
we are moving towards architecture where the visualization itself will have no knowledge about what aggregations produced the data it received.

so any usage of aggConfigs inside visualization should be avoided for easier migration.

also any usage of vis.API.anything should be avoided as well. The reason those were there was to hide angular usage, but we are migrating away from angular and idealy nowadays you should import whatever you need at the place you need it.

another concept that is about to change is how filters work. Currently visualizations know about aggregations used to get the data and can construct the filters based on that information. Since in the future visualizations won't know about aggregations they also won't be able to produce the filters. What they will rather do is just emit an event with the value that was clicked.

so probably it would be best to wai with this until 7.x. What was your target release for this ?

trevan commented:
@ppisljar, the only place that I added a usage of an aggConfig was in the metrics vis and it was following the same pattern there for the bucketAgg.  So if a migration has to be done, it would have to be done in any case in that code base.

Same with the vis.API.  I'm using the same api that is already used in that visualization.

All the other changes are just adding the createFilter function to the metrics aggs just like they exist in the bucket aggs.  Is the createFilter function planning to change?

As for when I would like this.  I would like it now.  I've been maintaining this functionality in my own fork because my users just expect to be able to do that.  Not sure why no one else has complained but any time I have a "big number" on my dashboard, they almost always try and click on it.

trevan commented:
If the changes to the metric visualization isn't ok, would it be alright if the changes to add the createFilter functions to the metric aggregations get merged?  If that gets in, I could then create my own metrics plugin and not have to maintain a fork.

tsullivan commented:
@ppisljar do you have any response to the last comment from @trevan ? 

Looks like this branch is out of date, and based on the conversation, the approach might not be want we want to adopt in the code.

Perhaps the goal should be brought up in a new enhancement issue, and we can talk about approaches before continuing the PR work?

elasticmachine commented:
Pinging @elastic/kibana-app

ppisljar commented:
We are now at a place where visualizations themself have no knowledge about what aggregations were used to produce the data they were given (the data might not come from aggregations at all), where this implementation wouldn't be possible.

@trevan  are you ok with closing this ?

trevan commented:
@ppisljar, I don't know why you think this isn't possible since it is just doing the same work as other visualizations.  But if you don't want this functionality, go ahead and close it.

ppisljar commented:
sorry for not being clear, I wanted to say that the current implementation wouldn't work, it should be updated to:
- visualization should emit event about what table cell was clicked (raw column id, row id, table)
- vis_filters should handle that correctly to generate metric filters

trevan commented:
I'm going to close it for now.
